src/Nethermind/Nethermind.Trie/TrieKey.cs

```diff
@@ -0,0 +1,529 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Trie;
+
+/// <summary>
+/// Represents a key used in a trie data structure. 
+/// This struct can hold a key in one or two byte array parts 
+/// for efficient concatenation and slicing operations.
+/// </summary>
+public readonly struct TrieKey
+{
+    // Predefined array instances for single-byte keys (0..15).
+    private readonly static byte[][] _singleBytes = [[0], [1], [2], [3], [4], [5], [6], [7], [8], [9], [10], [11], [12], [13], [14], [15]];
+    // Predefined TrieKey instances for single-byte keys (0..15).
+    private readonly static TrieKey[] _singleByteKeys = [new(_singleBytes[0]), new(_singleBytes[1]), new(_singleBytes[2]), new(_singleBytes[3]), new(_singleBytes[4]), new(_singleBytes[5]), new(_singleBytes[6]), new(_singleBytes[7]), new(_singleBytes[8]), new(_singleBytes[9]), new(_singleBytes[10]), new(_singleBytes[11]), new(_singleBytes[12]), new(_singleBytes[13]), new(_singleBytes[14]), new(_singleBytes[15])];
+
+    /// <summary>
+    /// An empty TrieKey instance (no bytes).
+    /// </summary>
+    public static TrieKey Empty { get; } = new(Array.Empty<byte>());
+
+    // These two byte arrays together represent the key. 
+    // If _keyPart1 is null, the entire key is in _keyPart0.
+    private readonly byte[]? _keyPart0;
+    private readonly byte[]? _keyPart1;
```

Comment:
Before I dive deeper, maybe we can use something like ReadOnlySequence<byte>?

Reply:
`ReadOnlySequence` would be much more fat. 2 objects + 2 ints.

Reply:
Also allocates if you read across the objects as a Span

Reply:
Also allocates if you read across the objects as a Span; works better for much larger sizes of data